File: mobile/ios/App/AnnouncementVoice/Sources/AnnouncementVoice/AnnouncementVoice.swift
import Foundation

/// How natural a voice sounds, worst first. Mirrors
/// `AVSpeechSynthesisVoiceQuality` without depending on it, so the ranking
/// builds and runs wherever Swift does.
enum VoiceQuality: Int, Comparable {
    case standard
    case enhanced
    case premium

    static func < (lhs: Self, rhs: Self) -> Bool { lhs.rawValue < rhs.rawValue }
}

/// An installed voice as the ranking reads it.
struct SpokenVoice {
    let identifier: String
    let language: String
    let quality: VoiceQuality
}

/// The regions the app has already chosen for the languages it speaks,
/// mirroring `dateLocale()` in `web/src/i18n/index.ts`.
private let regions = ["en": "en-GB", "sv": "sv-SE"]

/// The locale the announcements are spoken in. The web sends a bare language
/// and voices come per region, so an unqualified one is given the app's.
func announcementLocale(_ locale: String) -> String {
    regions[locale.lowercased()] ?? locale
}

/// The best-sounding installed voice for `locale`, or nil when none of them
/// beats what the synthesiser would have picked on its own.
///
/// Only an upgrade is ever reported, so a phone carrying nothing but the
/// standard voices keeps the voice it already announced in.
func bestVoice(for locale: String, among voices: [SpokenVoice]) -> SpokenVoice? {
    let wanted = announcementLocale(locale)
    let candidates = voices.filter {
        $0.quality > .standard && language(of: $0.language) == language(of: wanted)
    }
    // Ties are settled on the identifier so a phone offered two equals always
    // announces in the same one.
    return candidates.max { left, right in
        let ranks = (rank(left, for: wanted), rank(right, for: wanted))
        return ranks.0 == ranks.1 ? left.identifier > right.identifier : ranks.0 < ranks.1
    }
}

/// Where a voice stands, best last. The app's own region comes first: an
/// enhanced British voice is a better cue than a premium American one.
private func rank(_ voice: SpokenVoice, for locale: String) -> (region: Int, quality: Int) {
    (voice.language.lowercased() == locale.lowercased() ? 1 : 0, voice.quality.rawValue)
}

/// The language a locale names, with its region and its case dropped.
private func language(of locale: String) -> Substring {
    locale.lowercased().prefix { $0 != "-" }
}

/// A cue closed off with a full stop, which is what makes the synthesiser fall
/// away at the end of a phase instead of clipping it.
func announcementPhrase(_ instruction: String) -> String {
    let cue = instruction.trimmingCharacters(in: .whitespacesAndNewlines)
    guard let last = cue.last, !".!?,:;".contains(last) else { return cue }
    return cue + "."
}

/// The token a phrase holds a pause at, mirroring `pausePlaceholder` in
/// `web/src/native/announcementVoice.ts`.
let pausePlaceholder = "{pause}"

/// How long the synthesiser waits at one, in seconds.
let announcementPauseSeconds = 0.5

/// A phrase as the parts the synthesiser says it in, each closed off.
///
/// A full stop is the only pause a phrase can carry on its own, and no engine
/// gives one much of a beat. A phrase that wants one is split here, and the
/// wait is asked of `AVSpeechUtterance.preUtteranceDelay` instead.
func announcementParts(_ instruction: String) -> [String] {
    instruction.components(separatedBy: pausePlaceholder)
        .map(announcementPhrase)
        .filter { !$0.isEmpty }
}

File: web/src/native/announcementVoice.ts
/**
 * Which of the browser's voices says a cue, and how the cue is closed off.
 *
 * The phones rank the voices installed on them — `AnnouncementVoice.swift`,
 * the package the iOS app compiles in — because the one a synthesiser picks by
 * default is the flattest one it has. A browser ranks worse still: it says an
 * example in whatever voice its page language suggests, so a Swedish cue is
 * read out in English. This is the same ranking against what a browser will
 * tell us about a voice, which is its name, its language and whether it is
 * shipped or fetched.
 */

/**
 * The regions the app has already chosen for the languages it speaks,
 * mirroring `dateLocale()` in `web/src/i18n/index.ts`. Voices come per region,
 * and the web app knows only a language.
 */
const regions: Record<string, string> = { en: 'en-GB', sv: 'sv-SE' }

/** The locale a cue is spoken in: a bare language is given the app's region. */
export const announcementLocale = (locale: string): string =>
  regions[locale.toLowerCase()] ?? locale

/** The language a locale names, with its region and its case dropped. */
const language = (locale: string) => locale.toLowerCase().split('-')[0]

/**
 * How natural a voice sounds, worst first, read off the two things a browser
 * says about it.
 *
 * There is no quality field in the web API, so the name carries it: Safari and
 * the phones' WebViews mark theirs `(Enhanced)` or `(Premium)`. Chrome marks
 * nothing and ships the flat ones locally, fetching the natural ones — so a
 * voice the browser has to go and get is the better of two otherwise alike.
 */
const quality = (voice: SpeechSynthesisVoice): number => {
  const name = voice.name.toLowerCase()
  if (name.includes('premium')) return 2
  if (/enhanced|neural|natural/.test(name)) return 1
  return voice.localService ? 0 : 1
}

/** Where a voice stands, best last: the app's own region first, then quality. */
const rank = (voice: SpeechSynthesisVoice, locale: string): [number, number] => [
  voice.lang.toLowerCase().replace('_', '-') === locale.toLowerCase() ? 1 : 0,
  quality(voice),
]

/**
 * The best-sounding voice for `locale`, or nothing where none of them beats
 * what the browser would have picked on its own.
 *
 * Only an upgrade is ever reported, so a browser carrying nothing but its
 * ordinary voices keeps the one it already speaks in.
 */
export const bestVoice = (
  locale: string,
  voices: readonly SpeechSynthesisVoice[],
): SpeechSynthesisVoice | undefined => {
  const wanted = announcementLocale(locale)
  const candidates = voices.filter(
    (voice) => quality(voice) > 0 && language(voice.lang.replace('_', '-')) === language(wanted),
  )
  // Ties settle on the name, so a browser offered two equals always speaks in
  // the same one rather than a different one on every tap.
  return candidates.reduce<SpeechSynthesisVoice | undefined>((best, voice) => {
    if (!best) return voice
    const [region, sound] = rank(voice, wanted)
    const [bestRegion, bestSound] = rank(best, wanted)
    if (region !== bestRegion) return region > bestRegion ? voice : best
    if (sound !== bestSound) return sound > bestSound ? voice : best
    return voice.name < best.name ? voice : best
  }, undefined)
}

/**
 * A cue closed off with a full stop, which is what makes a synthesiser fall
 * away at the end of it instead of clipping the last word.
 */
export const announcementPhrase = (instruction: string): string => {
  const cue = instruction.trim()
  const last = cue.at(-1)
  return !last || '.!?,:;'.includes(last) ? cue : `${cue}.`
}
